### PR TITLE
add debug.saveBeforeRun setting

### LINKED ISSUE #114890: Running an untitled (unsaved) file, prompt to save each time
Issue Type: <b>Bug</b>

Hi,

I've been using VScode Insiders & exploration versions for about 2 years.
Only had one issue with them in the past, but now since a week or 2-3, each time i try to run some Powershell code without saving it, VScode prompts me for a 'save as'. I can get passed it by pressing escape but that gets old real quick.

See my gif for a visual representation of this issue.

VScode_saveAS

I have thse in both versions of Insiders & Exploration.

I do not have it in the up-to-date vanilla version so far.

VS Code version: Code - Exploration 1.53.0-exploration (0c9e2dadd986d0bc085f1c5352245531eef65e84, 2021-01-07T06:32:42.179Z)
OS version: Windows_NT x64 10.0.18363

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8250U CPU @ 1.60GHz (8 x 1800)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: enabled<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.85GB (5.75GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (7)</summary>

Extension|Author (truncated)|Version
---|---|---
bracket-pair-colorizer|Coe|1.0.61
gitlens|eam|11.1.3
notepadplusplus-keybindings|ms-|1.0.5
powershell-preview|ms-|2020.9.0
hide-gitignored|npx|1.1.0
material-icon-theme|PKi|4.5.0
gitblame|wad|6.0.2

(1 theme extensions excluded)

</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
openlogontheside:30221882
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492:30208929
wsl2promptcf:30219163
vstry914:30244316
pythonvsdeb440:30224570
unusedprompt:30219164
folderexplorer:30219166
openfilemenucf:30219169
pythonvsded773:30223139
pythonvspyt875:30247033
core-portspanel:30233467

```

</details>

### LINKED ISSUE #117080: Add a setting to not save before running a debug config
There exists a setting `task.saveBeforeRun` which controls whether all files are saved before running a task.
There should be a similar setting for when starting a debug session - `debug.saveBeforeRun`.

I ran into this and at first couldn't figure out why some of my files which I intentionally left unsaved, kept getting auto-saved.

I think this is the relevant code:
https://github.com/microsoft/vscode/blob/5d2c9bf29962533f37f34c36c58d4628a5c895a4/src/vs/workbench/contrib/debug/browser/debugService.ts#L282-L287

I think the same options of `always`/`prompt`/`never` should suffice, but we could also have `onlyCurrentFile` or something, or `onlyWorkspace` (since the files I had problems with weren't part of my workspace).

### PR DESCRIPTION
**Pull request by SnirBroshi**
This PR fixes #114890 and #117080

### CODE REVIEW

**Comment by NotWearingPants** [comment]
@isidorn I made the change you asked :)

Although I'm not sure what you mean by "EditorsInActiveGroup" so hopefully this is what you had in mind. The code saves all open non-untitled editors plus the current focused editor if it is untitled (this is the logic regardless of this PR), but "allEditorsInActiveGroup" implies non-focused untitled files being saved as well, which is not the case (and was also not requested by any issue I've seen).

I've enabled the "Allow edits and access to secrets by maintainers" checkbox if you want to modify my code tomorrow - if you want me to do any changes it will be at this time tomorrow. Thanks!

**Comment by isidorn** [comment]
This looks good.

Two minor changes still needed - I have commented in the code

File: src/vs/workbench/contrib/debug/browser/debugService.ts [hunk 1]
```diff
@@ -279,11 +279,17 @@ export class DebugService implements IDebugService {
 			// make sure to save all files and that the configuration is up to date
 			await this.extensionService.activateByEvent('onDebug');
 			if (!options?.parentSession) {
-				await this.editorService.saveAll();
-				const activeEditor = this.editorService.activeEditorPane;
-				if (activeEditor) {
-					// Make sure to save the active editor in case it is in untitled file it wont be saved as part of saveAll #111850
-					await this.editorService.save({ editor: activeEditor.input, groupId: activeEditor.group.id });
+				const saveBeforeStartConfig = this.configurationService.getValue('debug.saveBeforeStart');
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugService.ts:282]: use getValue<string> so the saveBeforeStartConfig is a string and not `any`

File: src/vs/workbench/contrib/debug/browser/debugService.ts [hunk 2]
```diff
@@ -279,11 +279,17 @@ export class DebugService implements IDebugService {
...
+
+				if (saveBeforeStartConfig !== 'none') {
+					await this.editorService.saveAll();
+					if (saveBeforeStartConfig === 'allEditorsInActiveGroup') {
+						const activeEditor = this.editorService.activeEditorPane;
+						if (activeEditor) {
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debugService.ts:288]: Please restructure this a bit such that you do not save all editors twice.

For example 
```
if (saveBeforeStartConfig === 'nonUntitledEditorsInActiveGroup') {
  .saveAll()
}
if (saveBeforeStartConfig === 'allEditorsInActiveGroup') {
  activeEditor...
}
```

File: src/vs/workbench/contrib/debug/browser/debug.contribution.ts [hunk 3]
```diff
@@ -489,6 +489,16 @@ function registerConfiguration(): void {
 				type: 'boolean',
 				description: nls.localize({ comment: ['This is the description for a setting'], key: 'showInlineBreakpointCandidates' }, "Controls whether inline breakpoints candidate decorations should be shown in the editor while debugging."),
 				default: true
+			},
+			'debug.saveBeforeStart': {
+				description: nls.localize('debug.saveBeforeStart', "Save all dirty editors before starting a debug session."),
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debug.contribution.ts:494]: Please update description to be better
`Controls what editors to save before starting a debug session.`

File: src/vs/workbench/contrib/debug/browser/debug.contribution.ts [hunk 4]
```diff
@@ -489,6 +489,16 @@ function registerConfiguration(): void {
...
+				enum: ['allEditorsInActiveGroup', 'nonUntitledEditorsInActiveGroup', 'none'],
+				enumDescriptions: [
+					nls.localize('debug.saveBeforeStart.allEditorsInActiveGroup', "Save all editors before starting a debug session."),
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/browser/debug.contribution.ts:497]: Please add `in active group` in description

**Review by isidorn — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/debug/browser/debugService.ts (hunk 2 above)
Comment (NotWearingPants) [reply src/vs/workbench/contrib/debug/browser/debugService.ts:288]: I don't see how the second branch in your comment is supposed to work. Judging by the previous code, `saveAll` is still needed to save everything in both cases, and the `allEditorsInActiveGroup` case just adds another save to the current one.

This is the logic you wrote in cf8ed37206e161a30151defa4fc9ea58c32335bc, where did I introduce a double save?

Comment (isidorn) [reply src/vs/workbench/contrib/debug/browser/debugService.ts:288]: You are correct, I was mistaken. Please ignore my comment.

**Review by isidorn — APPROVED** [review]

**Comment by isidorn** [comment]
This works nicely. Thanks for working on this. Merging in ☀️

**Pull request merged**